### PR TITLE
Crude custom graphics for machine elements

### PR DESCRIPTION
**Pull request by dlbeswick**
See the graphics drawn on the "kick" and "distort" machines below:

Screenshot from 2021-06-19 23-41-47-1

I wanted to be able to visualise things like my distort function and my kick's volume envelope, so I added this in. Very open to discussion on this, it's ugly and more of a starting point. For example, maybe this could be shown via a "drawer" the pops open at the bottom of the machine, rather than drawing over the standard gfx.

This signal-based approach might also point the way to things such as machine-specific formatting of parameter values, like Buzz used to have. I'm conscious of trying to avoid hard dependencies between the UI and machine.

Have a look here to see it in action in a machine: https://github.com/dlbeswick/bt-edp-distort/blob/master/src/machine.c

It works via three signals:

1. Either the UI or machine can emit "bt-gfx-invalidated" to ask for a redraw. The machine will emit this when parameters change, for instance.
1. The UI connects to "bt-gfx-invalidated" and will shortly afterwards emit "bt-gfx-request" to the machine, to ask for the graphics.
1. The machine in response will draw an ABGR image and emit "bt-gfx-present" back to the UI, passing the width, height and image data.

This could be pared down to two signals if I could figure out the best way to pass the width, height and image data as a return value from "bt-gfx-request". Struct? GObject? I don't want to have to link with any UI/edit-specific data structures, although headers are obviously ok.

It's important to make sure the image gets set in the main thread, and a glib timer is used to do this. The "invalidate" logic is also used so that many property updates (i.e. when the song starts) only resolve to a single gfx update call.

### CODE REVIEW

**Comment by dlbeswick** [comment]
I took another pass over it and removed the need for the third signal by return a pointer to a struct from "bt-gfx-request". 

I've left both commits there for now, I'll squash them for you if you'd like to pull.

File: src/ui/edit/machine-canvas-item.c [hunk 1]
```diff
@@ -194,6 +200,58 @@ update_machine_graphics (BtMachineCanvasItem * self)
...
+  // 1. It makes sure that the image update happens on the main thread.
+  //    Clutter doesn't seem to like requests coming from another thread.
+  //    This will likely be called from the Gstreaming processing thread.
+  // 2. It aggregates many invalidating requests (i.e. from multiple machine
+  //    property updates) into a single "update gfx" call.
+  g_mutex_lock(&self->priv->custom_gfx_lock);
+  if (!self->priv->custom_gfx_timer)
+    self->priv->custom_gfx_timer =
+        g_timeout_add(5, G_SOURCE_FUNC(custom_gfx_request_after_invalidation),
+            self);
```
Comment (ensonic) [inline src/ui/edit/machine-canvas-item.c:251]: I think you can use g_idle_add for this

**Review by ensonic — COMMENTED** [review]
I really like the idea. Maybe its also time for a little facelift on the machine icons to remove some bevels (like put the title on a top frame). I wonder If we could rearrange the things a bit to avoid the need to paint over everything (e.g. show these graphics or the mute/solo buttons). But in principle have to get this in as with the comment I placed and then we can iterate.

File: src/ui/edit/machine-canvas-item.c (hunk 1 above)
Comment (dlbeswick) [reply src/ui/edit/machine-canvas-item.c:251]: Ok, great!

I didn't even do my maths properly here and so didn't set the timeout value properly, but my intention here was to limit the gfx refreshes to roughly 60fps to avoid wasting CPU. The timeout value should be 16, not 5.

Do you think that using the idle function will use more CPU in practice?

Comment (ensonic) [reply src/ui/edit/machine-canvas-item.c:251]: Actually, you are right. A timer makes more sense here. You will periodically request the machine to update the image buffer.

File: src/lib/gst/ui.h [hunk 2]
```diff
@@ -0,0 +1,35 @@
+/* Buzztrax
+ * Copyright (C) 2015 Stefan Sauer <ensonic@users.sf.net>
```
Comment (ensonic) [inline src/lib/gst/ui.h:2]: Please use you name here :)

**Comment by ensonic** [comment]
Just food for thought. I also wondered if we could have more graphs in the machine UI (e.g. for oscillators, filters, etc). I wonder if we want to make this an interface for the machine to implement. Then the UI could 
  * query the machien for how many previews it supports
  * the request signal could be an ordinary method
 
The part that I did not really have through thru is how the machine would tell that a gfx-buffer e.g. relates to properties x,y,z. Lets maybe brainstorm a little more on that in the chat.

File: src/lib/gst/ui.h (hunk 2 above)
Comment (dlbeswick) [reply src/lib/gst/ui.h:2]: Whoops! Thanks

File: src/ui/edit/machine-canvas-item.c (hunk 1 above)
Comment (dlbeswick) [reply src/ui/edit/machine-canvas-item.c:251]: Not quite -- the update aren't periodic. The graphics are only flagged for an update when the machine's "bt-gfx-invalidated" signal is emitted. Currently this happens when the machine is first assigned to the BtMachineCanvasItem, and then whenever the machine itself wants to request an update.

I imagine the machine will almost always be emitting "bt-gfx-invalidated" when one of its properties is updated. This is what my current implementation is doing. However, a machine could also request updates periodically if it wants to do some "realtime" display. So there's some flexibility there.

You see bursts of updates when the song starts, and also when the user is dragging sliders back and forth. So I was thinking this is the reason for the timer over the idle callback -- I think the idle callback could easily end up making more than 60fps of graphics updates during these bursts, even though it wasn't updating on every call. However, the timer would limit the max update rate. But I might have the wrong idea about how often/quickly the idle callback gets called.

**Comment by dlbeswick** [comment]
It would be great to provide a means for machines to easily draw graphs for some common properties. This goes along with something else I was thinking; attaching some more semantic information to the properties. I.e. a filter's cutoff value is semantically a "frequency value". The property is defined as a linear 0 to 22khz, but the slider in Buzztrax applies logarithmic scaling for convenience for the user. The machine author also doesn't need to repeatedly apply this kind of scaling in his "property set" code, etc. With more of this semantic information, you could also pick out more candidates for visualisation. 

On another point, this code would already also be better implemented as an interface, ideally. The reason that I chose signals over an interface is because I have the impression that the machine will have to link with a Buzztrax library to use an interface, which could make it awkward to use as a regular GStreamer machine. Is this one of your design goals -- to have Buzztrax machines also usable in GStreamer pipelines? I assumed it was.

So in short, if you'd like me to re-work this as an interface then I'm more than happy to do that, I'd just need a few pointers still being a Glib novice! Will talk more in chat.

File: src/ui/edit/machine-canvas-item.c (hunk 1 above)
Comment (dlbeswick) [reply src/ui/edit/machine-canvas-item.c:251]: Fixed that timer value and squashed commits, btw.

Comment (ensonic) [reply src/ui/edit/machine-canvas-item.c:251]: The bursts of updates are because when you set the song to play, the CPU quickly spikes to flood the pipeline with buffers (preroll). Once all elements in a pipeline have buffers, the audio-sinks and the queue's stall upstream branches to only send more data as it is needed.
I did not thought of that when doing the review (have been away from the code for too long). Anyway please take a look at e.g. gstreamer Level element (or Spectrum) and how it is used in buzztrax. The data that is emitted has the play-back time and I use a g_timeout_add to defer drawing to the point in time when the audio is played. This is especially noticable for sources with lots of fx.
Maybe we can do the same? Emit the data as a GstMessage on the bus and include a timestamp? Another option is an GstEvent that is emitted by the sink as a GstMessage (https://gstreamer.freedesktop.org/documentation/gstreamer/gstevent.html?gi-language=c#gst_event_new_sink_message). This has the advantage that Gsteamer will handle the syncronisation, but there is a little overhead on passing the data along and if you have multiple sinks (e.g. when rendering a wav to disk) the data is copied and the app needs to deduplicate the Message.

Finally, regarding he interface - I think we don't need to worry right now (if the GstMessage works), but the interface can be optional. The element would still work as a normal GstElement.

**Comment by ensonic** [comment]
For the ideas on how to improve the UIs (and also general future development), maybe we can meet in gitter.im - still need to be able to catch you online though.

**Comment by dlbeswick** [comment]
Sorry, I know I keep dodging chat. My free time to work on this is really irregular, but I'll try to make some time to chat.

I currently plan re-work this to use an interface in the near future; not sure why I was so hung up on Buzztrax lib dependencies as these are needed for audiosynth machines anyhow.

File: src/ui/edit/machine-canvas-item.c (hunk 1 above)
Comment (dlbeswick) [reply src/ui/edit/machine-canvas-item.c:251]: Thank you, I'll take a look at Spectrum to see how it works. One difference I can immediately think of is that unlike spectrum analysis, the generation of the sound isn't usually the "event" leading to the update. In my mind the most common trigger is a UI interaction with a parameter, and the next most common a pattern position change. This applies to the kinds of visualisations I'm using now, which are just parameter-based. 

Although, there would be other classes of visualisations that would be more like the spectrum case, i.e. if an author wanted to visualise k-rate or a-rate internal values of the machine (I'm using some CSound terms here,) so it might might sense to collapse everything to this case.

I'm very happy to let this one bake and iterate on it until we have a good approach to work with.

**Comment by ensonic** [comment]
No rush. Also I think the event suggestion I made could help to simplify it.

**Comment by dlbeswick** [comment]
Hello Stefan, apologies for this, I don't really like to leave things half-finished, I just have too many hobbies.

I added the interface, which hasn't made much practical difference, but the code is a bit safer and I think it's a gain in the way it points to how similar features could be added in future. Another benefit is that GObject interfaces can define signals on their implementing classes automatically; I found this a bit surprising, but it's useful.

I have kept the interface design as granular as possible so machine authors can pick and choose which features they would like to support by implementing interfaces. I think this is a good way to go.

The "invalidate" signal has remained, and I think this makes sense because this is a form of communication "up" the hierarchy from the "child" machine to the "parent" UI.

So, to anyone extending this I would say:

* Define interface methods for "push"-style communications from UI to machine, i.e. UI asks machine for property display name.
* Define signals for the (more rare) case of machine to UI communication.

I haven't looked at using GstEvent instead of signals, but I can. My only question is: will these GstEvents still get propagated if the pipeline isn't running? I suspect it won't, but that's not the best for this case because you ideally want to see these kinds of graphics changing as you tweak parameter values.

Any and all feedback welcome as always, hope I wrote the docs correctly in particular. And if you don't get to it soon then no bother, it's been a while between updates from me!

**Comment by dlbeswick** [comment]
Also wanted to make sure the naming of the signals and classes were ok; I aimed to situate it in the GstBt module with some kind of "Ui" subcategory and avoid any clashes with signal names.

**Comment by ensonic** [comment]
Thanks a lot for the followup! I merged the change. I still wonder if we maybe want to rethink the layout of the icons, so that we don't have to draw over. Maybe we can also offset the overlay a bit and account for the shadow to make it fit.

Another idea could be an option to enable/disable it and/or an hotkey to toggle it. If we toggle with an hotkey we could either show the level meters and the mute/bypass/solo buttons or the custom gfx.

Let me know if you have some thoughts here.
